reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,383 @@
+            });
+            iterate(triggeredCalls).filter(call -> !choices.containsKey(call)).forEachRemaining(nextPendingKeys::add);
+            ComponentPlan newPlan = componentPlanSearch(root, nextPendingKeys, choices);
+
+            double newPlanCost = newPlan.cost(root);
+            if (bestPlan == null || newPlanCost < bestPlanCost) {
+                bestPlan = newPlan;
+                bestPlanCost = newPlanCost;
+            }
+
+            choices.remove(mode);
+        }
+        assert bestPlan != null;
+        return bestPlan;
+    }
+
+    private void initialiseOrderingDependencies(ConjunctionNode conjunctionNode, List<Resolvable<?>> ordering, Set<Variable> inputBounds) {
+        Set<Variable> currentBounds = new HashSet<>(inputBounds);
+        for (Resolvable<?> resolvable : ordering) {
+            Set<Variable> resolvableBounds = iterate(consideredVariables(resolvable)).filter(currentBounds::contains).toSet();
+            initialiseDependencies(conjunctionNode, resolvable, resolvableBounds);
+            if (!resolvable.isNegated()) currentBounds.addAll(consideredVariables(resolvable));
+        }
+    }
+
+    private void initialiseDependencies(ConjunctionNode conjunctionNode, Resolvable<?> resolvable, Set<Variable> resolvableBounds) {
+        if (resolvable.isConcludable()) {
+            Set<ResolvableConjunction> cyclicDependencies = conjunctionNode.cyclicDependencies(resolvable.asConcludable());
+            for (CallMode callMode : triggeredCalls(resolvable.asConcludable(), resolvableBounds, Optional.empty())) {
+                recursivelyGenerateOrderingChoices(callMode);
+                if (!cyclicDependencies.contains(callMode.conjunction)) {
+                    plan(callMode); // Acyclic dependencies can be fully planned
+                }
+            }
+        } else if (resolvable.isNegated()) {
+            iterate(resolvable.asNegated().disjunction().conjunctions()).forEachRemaining(conjunction -> {
+                CallMode callMode = new CallMode(conjunction, resolvableBounds);
+                recursivelyGenerateOrderingChoices(callMode);
+                plan(callMode);
+            });
+        }
+    }
+
+    private void recursivelyGenerateOrderingChoices(CallMode callMode) {
+        if (!candidateOrderings.containsKey(callMode)) {
+            candidateOrderings.put(callMode, null); // Guard
+            ConjunctionNode conjunctionNode = conjunctionGraph.conjunctionNode(callMode.conjunction);
+            answerCountEstimator.buildConjunctionModel(callMode.conjunction);
+
+            // Two orderings for the same CallMode with identical cyclic-concludable modes are interchangeable in the component-plan
+            //      -> We only need to keep the cheaper one.
+            Map<Set<Pair<Concludable, Set<Variable>>>, OrderingChoice> bestChoice = new HashMap<>();
```

Comment:
you really love "best" haha!

The point of this method, if we use the new naming and keep it verbose for now, is to `generateCallGraphPlanChoices`. So really, the main point of this variable is not to store the "best" choice right? That's a detail of resolving two interchangable plans. The main thing is to actually create the choices. So i'd call this `planChoices`, and move the comment you have above it to the `if` statement below